File: selection-sort/selection_sort.py
def find_minimum(arr):
    """Returns the index of the minimum value in an array
    """
    min_val = arr[0]
    min_index = 0
    for i in range(1, len(arr)):
        if arr[i] < min_val:
            min_val = arr[i]
            min_index = i
    return min_index


def selection_sort(arr):
    """Sort an array in ascending order using selection sort.

    Uses arr.pop to produce unsorted subarray instead of tracking index of
    sorted items.
    """
    sorted_arr = []
    for i in range(len(arr)):
        min_index = find_minimum(arr)
        sorted_arr.append(arr.pop(min_index))
    return sorted_arr


array = [22, 11, 99, 88, 9, 7, 42]
print("Unsorted = %s" % array)
array = selection_sort(array)
assert(array == [7, 9, 11, 22, 42, 88, 99])
print("Sorted = %s" % array)

array = [-3, 10, -8, 11]
print("Unsorted = %s" % array)
array = selection_sort(array)
assert(array == [-8, -3, 10, 11])
print("Sorted = %s" % array)

File: selection-sort/selectionSort.js
/**
 * Swaps two items position in an array.
 * @param {array} array array to mutate 
 * @param {number} firstIndex Index of first item to swap.
 * @param {number} secondIndex Index of second item to swap.
 */
let swap = function (array, firstIndex, secondIndex) {
    let tmp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = tmp;
};

/**
 * Finds the index of the minimum value in a subarray.
 * @param {array} array entire array containing the subarray
 * @param {number} startIndex Index at which subarray begins
 * @returns {number} Index of minimum value within subarray
 */
let indexOfMinimum = function (array, startIndex) {

    let minValue = array[startIndex];
    let minIndex = startIndex;

    for (let i = minIndex + 1; i < array.length; i++) {
        if (array[i] < minValue) {
            minIndex = i;
            minValue = array[i];
        }
    }
    return minIndex;
};

/**
 * Sort an array into ascending order using selection sort.
 * @param {array} array array to sort 
 */
let selectionSort = function (array) {
    let min;
    for (let i = 0; i < array.length; i++) {
        // Get index of min value from unsorted subarray
        min = indexOfMinimum(array, i);
        // Swap into correct place
        swap(array, i, min);
    }
};
/**
 * Compares contents and order of simple, 1d arrays
 * @param {array} arr1 
 * @param {array} arr2 
 * @return {bool} Whether arrays are equal
 */
let arrayAreEqual = function (arr1, arr2) {
    return arr1.lenth === arr2.lenth && arr1.reduce((a, b, i) => a && arr1[i] === arr2[i], true)
}
let array = [22, 11, 99, 88, 9, 7, 42];
selectionSort(array);
console.assert(true === arrayAreEqual(array, [7, 9, 11, 22, 42, 88, 99]));

let array = [-3, 10, -8, 11];
selectionSort(array);
console.assert(true === arrayAreEqual(array, [-8, -3, 10, 11]));